src/nwb_conversion_tools/tools/spikeinterface/spikeinterface.py

```diff
@@ -1065,38 +1054,50 @@ def add_units(
-        - 'units' will save it to the official NWBFile.Units position; recommended only for the final form of the data.
-        - 'processing' will save it to the processing module to serve as a historical provenance for the official table.
-    units_name : str (optional, defaults to 'units')
-        The name of the units table. If write_as=='units', then units_name must also be 'units'.
-    units_description : str (optional)
-        Text description of the sorting table; recommended to included parameters of sorting method, curation, etc.
+    write_in_processing_module: bool (optional, defaults to False)
+        How to save the units table in the nwb file.
+        - True will save it to the processing module to serve as a historical provenance for the official table.
+        - False will save it to the official NWBFile.Units position; recommended only for the final form of the data.
+    units_table_name : str (optional, defaults to 'units')
+        The name of the units table. If write_as=='units', then units_table_name must also be 'units'.
+    unit_table_description : str (optional)
+        Text description of the units table; it is recommended to include information such as the sorting method,
+        curation steps, etc.
     write_waveforms : bool (optional, defaults to false)
         if True and sorting is a spikeextractors SortingExtractor object then waveforms are added to the units table
         after writing.
     """
+    if not isinstance(nwbfile, pynwb.NWBFile):
+        raise TypeError(f"nwbfile type should be an instance of pynwb.NWBFile but got {type(nwbfile)}")
+
     if isinstance(sorting, SortingExtractor):
-        sampling_frequency = sorting.get_sampling_frequency()
-        if sampling_frequency is None:
-            raise ValueError("Writing a SortingExtractor to an NWBFile requires a known sampling frequency!")
         checked_sorting = OldToNewSorting(oldapi_sorting_extractor=sorting)
     else:
         checked_sorting = sorting
-    unit_ids = checked_sorting.get_unit_ids()
+
     sampling_frequency = checked_sorting.get_sampling_frequency()
     if sampling_frequency is None:
         raise ValueError("Writing a SortingExtractor to an NWBFile requires a known sampling frequency!")
-    assert write_as in [
-        "units",
-        "processing",
-    ], f"Argument write_as ({write_as}) should be one of 'units' or 'processing'!"
-    if write_as == "units":
-        assert units_name == "units", "When writing to the nwbfile.units table, the name of the table must be 'units'!"
+
+    if not write_in_processing_module:
+        units_table_name == "units"
+
+    if write_in_processing_module:
+        ecephys_mod = get_module(
+            nwbfile=nwbfile,
+            name="ecephys",
+            description="Intermediate data from extracellular electrophysiology recordings, e.g., LFP.",
+        )
+        if units_table_name not in nwbfile.processing:
```

Comment:
Does this recurse across all the modules of `nwbfile.processing`? Shouldn't it check if it's in the `ecephys_mod`?

Reply:
You are correct. It should test whether it is in ecephys_mod.data_interfaces

I changed it. I added a test for writing in processing as well.